File: app/src/main/java/org/take365/Components/ProgressRequestBody.java
package org.take365.Components;

import android.os.Handler;
import android.os.Looper;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.BufferedSink;

/**
 * Created by divere on 31/10/2016.
 */

public class ProgressRequestBody extends RequestBody {
    private File mFile;
    private InputStream mInputStream;
    private String mPath;
    private UploadCallbacks mListener;

    private static final int DEFAULT_BUFFER_SIZE = 2048;

    public interface UploadCallbacks {
        void onProgressUpdate(int percentage);
        void onError();
        void onFinish();
    }

    public ProgressRequestBody(final File file, final  UploadCallbacks listener) {
        mFile = file;
        mListener = listener;
    }

    public ProgressRequestBody(final InputStream inputStream, final  UploadCallbacks listener) {
        mInputStream = inputStream;
        mListener = listener;
    }

    @Override
    public MediaType contentType() {
        // i want to upload only images
        return MediaType.parse("image/*");
    }

    @Override
    public long contentLength() throws IOException {
        if(mFile != null){
            return mFile.length();
        }

        if(mInputStream != null) {
            return mInputStream.available();
        }

        return 0;
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {

        byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
        InputStream in = null;
        long fileLength = 0;
        if(mFile != null) {
            fileLength = mFile.length();
            in = new FileInputStream(mFile);
        }
        if(mInputStream != null) {
            fileLength = mInputStream.available();
            in = mInputStream;
        }

        if(in == null) {
            return;
        }

        long uploaded = 0;

        try {
            int read;
            Handler handler = new Handler(Looper.getMainLooper());
            while ((read = in.read(buffer)) != -1) {

                // update progress on UI thread
                handler.post(new ProgressUpdater(uploaded, fileLength));

                uploaded += read;
                sink.write(buffer, 0, read);
            }
        } finally {
            in.close();
        }
    }

    private class ProgressUpdater implements Runnable {
        private long mUploaded;
        private long mTotal;
        public ProgressUpdater(long uploaded, long total) {
            mUploaded = uploaded;
            mTotal = total;
        }

        @Override
        public void run() {
            mListener.onProgressUpdate((int)(100 * mUploaded / mTotal));
        }
    }
}

File: app/src/main/java/org/take365/Components/ProgressRequestBody.kt
package org.take365.Components

import android.os.Handler
import android.os.Looper

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream

import okhttp3.MediaType
import okhttp3.RequestBody
import okio.BufferedSink

/**
 * Created by divere on 31/10/2016.
 */

class ProgressRequestBody : RequestBody {
    private var mFile: File? = null
    private var mInputStream: InputStream? = null
    private val mPath: String? = null
    private var mListener: UploadCallbacks? = null

    interface UploadCallbacks {
        fun onProgressUpdate(percentage: Int)
        fun onError()
        fun onFinish()
    }

    constructor(file: File, listener: UploadCallbacks) {
        mFile = file
        mListener = listener
    }

    constructor(inputStream: InputStream, listener: UploadCallbacks) {
        mInputStream = inputStream
        mListener = listener
    }

    override fun contentType(): MediaType {
        // i want to upload only images
        return MediaType.parse("image/*")
    }

    @Throws(IOException::class)
    override fun contentLength(): Long {
        return mFile?.length() ?: (mInputStream?.available()?.toLong() ?: 0)

    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {

        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        var `in`: InputStream? = null
        var fileLength: Long = 0
        if (mFile != null) {
            fileLength = mFile!!.length()
            `in` = FileInputStream(mFile)
        }
        if (mInputStream != null) {
            fileLength = mInputStream!!.available().toLong()
            `in` = mInputStream
        }

        if (`in` == null) {
            return
        }

        var uploaded: Long = 0

        try {
            val handler = Handler(Looper.getMainLooper())
            while (true) {
                val read = `in`.read(buffer)
                if (read == -1) return
                // update progress on UI thread
                handler.post(ProgressUpdater(uploaded, fileLength))

                uploaded += read.toLong()
                sink.write(buffer, 0, read)
            }
        } finally {
            `in`.close()
        }
    }

    private inner class ProgressUpdater(private val mUploaded: Long, private val mTotal: Long) : Runnable {

        override fun run() {
            mListener!!.onProgressUpdate((100 * mUploaded / mTotal).toInt())
        }
    }

    companion object {

        private val DEFAULT_BUFFER_SIZE = 2048
    }
}
